packages/frontend/workspace/src/list/index.ts

```diff
@@ -0,0 +1,292 @@
+   */
+  async create(
+    flavour: WorkspaceFlavour,
+    initial: (
+      workspace: BlockSuiteWorkspace,
+      blobStorage: BlobStorage
+    ) => Promise<void>
+  ) {
+    const provider = this.providers.find(x => x.name === flavour);
+    if (!provider) {
+      throw new Error(`Unknown workspace flavour: ${flavour}`);
+    }
+    const id = await provider.create(initial);
+    const metadata = {
+      id,
+      flavour,
+    };
+    // update workspace list
+    this.status = this.addWorkspace(this.status, metadata);
+    return id;
+  }
+
+  /**
+   * delete workspace
+   * @param workspaceMetadata
+   */
+  async delete(workspaceMetadata: WorkspaceMetadata) {
+    logger.info(
+      `delete workspace [${workspaceMetadata.flavour}] ${workspaceMetadata.id}`
+    );
+    const provider = this.providers.find(
+      x => x.name === workspaceMetadata.flavour
+    );
+    if (!provider) {
+      throw new Error(
+        `Unknown workspace flavour: ${workspaceMetadata.flavour}`
+      );
+    }
+    await provider.delete(workspaceMetadata.id);
+
+    // delete workspace from list
+    this.status = this.deleteWorkspace(this.status, workspaceMetadata);
+  }
+
+  /**
+   * add workspace to list
+   */
+  private addWorkspace(
+    status: WorkspaceListStatus,
+    workspaceMetadata: WorkspaceMetadata
+  ) {
+    if (status.workspaceList.some(x => x.id === workspaceMetadata.id)) {
+      return status;
+    }
+    return {
+      ...status,
+      workspaceList: status.workspaceList.concat(workspaceMetadata),
+    };
+  }
+
+  /**
+   * delete workspace from list
+   */
+  private deleteWorkspace(
+    status: WorkspaceListStatus,
+    workspaceMetadata: WorkspaceMetadata
+  ) {
+    if (!status.workspaceList.some(x => x.id === workspaceMetadata.id)) {
+      return status;
+    }
+    return {
+      ...status,
+      workspaceList: status.workspaceList.filter(
+        x => x.id !== workspaceMetadata.id
+      ),
+    };
+  }
+
+  /**
+   * callback for subscribe workspaces list
+   */
+  private handleWorkspaceChange(changed: {
+    added?: WorkspaceMetadata[];
+    deleted?: WorkspaceMetadata[];
+  }) {
+    let status = this.status;
+
+    for (const added of changed.added ?? []) {
+      status = this.addWorkspace(status, added);
+    }
+    for (const deleted of changed.deleted ?? []) {
+      status = this.deleteWorkspace(status, deleted);
+    }
+
+    this.status = status;
+  }
+
+  /**
+   * start first load workspace list
+   */
+  private startLoad() {
+    Promise.allSettled(
+      this.providers.map(async provider => {
+        const unsubscribe = provider.subscribe(changed => {
+          this.handleWorkspaceChange(changed);
+        });
+        if (this.abortController.signal.aborted) {
+          unsubscribe();
+          return;
+        }
+        this.abortController.signal.addEventListener('abort', () => {
+          unsubscribe();
+        });
+        try {
+          const list = await provider.getList();
```

Comment:
I guess we may need to refresh periodically the list if the user has logged in.

Reply:
https://github.com/toeverything/AFFiNE/blob/4470eb39ccf53e0b8735fad7f2c783af4ada74ea/packages/frontend/workspace/src/list/index.ts#L227

In fact, WorkspaceList subscribes to the workspace list update from the provider. How to implement subscription updates is implemented by the provider.

Currently, both `local` and `cloud` providers listen to events from other tabs to rescan the list, so if the user manually created a workspace, it will be automatically sync between tabs.

But there is currently no way to know if the workspace list changes come from the server.

If you want to set up a scheduled refresh, I have no idea about this interval time.

Reply:
How about revalidate the workspace list when opening the list like SWR?

Reply:
I can add a revalidate function